### PR TITLE
Fix private module loophole in the 'private type in public item' check

### LINKED ISSUE #22261: private type in public API of non-pub mod is allowed
While reviewing the list of active feature gates and double-checking the behavior of `visible_private_types`, I found that the check is not working for e.g. the program below:

```rust
mod a {
    #[derive(Debug)]
    struct S { x: i32 }
    pub fn foo() -> S { S { x: 3 } }
    impl S { }
}

mod b {
    use a;

    #[cfg(expected)]
    pub fn call_foo() {
        let s: a::S = a::foo();
        println!("{:?}", s);
    }

    #[cfg(not(expected))]
    pub fn call_foo() {
        let s = a::foo();
        println!("{:?}", s);
    }
}

fn main() {
    b::call_foo();
}
```

In the playpen, the above compiles and runs.  I expected to see an error due to the type of `a::foo` returning the private type `a::S`.

### PR DESCRIPTION
**Pull request by nrc**
Closes #22261

r? @nikomatsakis 

(+ a new test coming soon...)

### CODE REVIEW

**Comment by nrc** [comment]
Now with test

**Comment by nrc** [comment]
testing, ignore this

triage: I-nominated (1.0 beta)

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage: I-nominated (1.0 beta)

**Comment by nrc** [comment]
triage: P-high

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage P-high

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage P-high

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage P-high

**Comment by nrc** [comment]
triage P-high

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triage P-high

**Comment by nrc** [comment]
rebased

File: src/librustc_privacy/lib.rs [hunk 1]
```diff
@@ -1321,7 +1321,18 @@ impl<'a, 'tcx, 'v> Visitor<'v> for VisiblePrivateTypesVisitor<'a, 'tcx> {
                     match *trait_ref {
                         None => {
                             for impl_item in impl_items {
-                                visit::walk_impl_item(self, impl_item);
+                                match impl_item.node {
+                                    ast::MethodImplItem(..)
+                                        if self.exported_items.contains(&impl_item.id) ||
```
Comment (nikomatsakis) [inline src/librustc_privacy/lib.rs]: can you maybe add a comment about what's going on here? I'm not sure I understand, though I presume this is related to the rules about privacy and impls?

File: src/librustc_privacy/lib.rs [hunk 2]
```diff
@@ -1321,7 +1321,18 @@ impl<'a, 'tcx, 'v> Visitor<'v> for VisiblePrivateTypesVisitor<'a, 'tcx> {
...
+                                           impl_item.vis == ast::Public =>
```
Comment (nikomatsakis) [inline src/librustc_privacy/lib.rs]: maybe we need an accessor for this compound condition (exported_items.contains || vis == pub), it appears also below...?

File: src/librustc_privacy/lib.rs [hunk 3]
```diff
@@ -1360,7 +1371,8 @@ impl<'a, 'tcx, 'v> Visitor<'v> for VisiblePrivateTypesVisitor<'a, 'tcx> {
                         match impl_item.node {
                             ast::MethodImplItem(ref sig, _) => {
                                 if sig.explicit_self.node == ast::SelfStatic &&
-                                   self.exported_items.contains(&impl_item.id) {
+                                        (self.exported_items.contains(&impl_item.id) ||
+                                         impl_item.vis == ast::Public) {
```
Comment (nikomatsakis) [inline src/librustc_privacy/lib.rs]: ...here.

File: src/librustc_privacy/lib.rs [hunk 4]
```diff
@@ -1381,15 +1393,19 @@ impl<'a, 'tcx, 'v> Visitor<'v> for VisiblePrivateTypesVisitor<'a, 'tcx> {
             ast::ItemTy(..) => return,
 
             // not at all public, so we don't care
-            _ if !self.exported_items.contains(&item.id) => return,
+            _ if !(self.exported_items.contains(&item.id) ||
+                   item.vis == ast::Visibility::Public) => {
```
Comment (nikomatsakis) [inline src/librustc_privacy/lib.rs]: ...and here.

**Comment by nikomatsakis** [comment]
r+ modulo nits

**Pull request merged**